### ISSUE TITLE
Semantics of prune retention policies

### ISSUE THREAD

**Issue by losvald** [borgbackup/borg#602]
I'm a bit confused by the behavior of --keep-hourly (and similar options). Would somebody mind explaining it a bit or ideally like "borg help retention" with examples?

Here is my first dry run of prune with `-H 2` (output is chronologically sorted):

	Keeping archive: 2016-01-24-same                      Sun, 2016-01-24 22:57:03
	Would prune:     2016-01-24                           Sun, 2016-01-24 22:54:12
	Keeping archive: 2016-01-21-updated2                  Thu, 2016-01-21 20:38:23
	Would prune:     2016-01-21-updated                   Thu, 2016-01-21 20:34:50
	Would prune:     tmp                                  Thu, 2016-01-21 17:44:08
	Would prune:     2016-01-21                           Thu, 2016-01-21 16:32:57

And the second one with `-H 3`:

	Keeping archive: 2016-01-24-same                      Sun, 2016-01-24 22:57:03
	Would prune:     2016-01-24                           Sun, 2016-01-24 22:54:12
	Keeping archive: 2016-01-21-updated2                  Thu, 2016-01-21 20:38:23
	Would prune:     2016-01-21-updated                   Thu, 2016-01-21 20:34:50
	Keeping archive: tmp                                  Thu, 2016-01-21 17:44:08
	Would prune:     2016-01-21                           Thu, 2016-01-21 16:32:57

Based on these, I inferred *not more than 1 backup in the same hour*, but does the day matter? Let's find out after making a few more backups and running with `-H 6`...

	Keeping archive: 2016-01-25-same                      Mon, 2016-01-25 16:03:51
	Keeping archive: 2016-01-25                           Mon, 2016-01-25 15:32:12
	Keeping archive: 2016-01-24-same                      Sun, 2016-01-24 22:57:03
	Would prune:     2016-01-24                           Sun, 2016-01-24 22:54:12
	Keeping archive: 2016-01-21-updated2                  Thu, 2016-01-21 20:38:23
	Would prune:     2016-01-21-updated                   Thu, 2016-01-21 20:34:50
	Keeping archive: tmp                                  Thu, 2016-01-21 17:44:08
	Keeping archive: 2016-01-21                           Thu, 2016-01-21 16:32:57

It seems it does; it kept 2016-01-21 16:32. But what would happen if there was an older entry that day, say 2016-01-21 14:00, would it pick that one instead?

IMHO, this is not analogous to the definition of --keep-daily from the doc:

> As an example, "-d 7" means to keep the latest backup on each day for 7 days.
> Days without backups do not count towards the total.

because in the case of `-H 3` the backups at 20:38 and 17:44 were kept
do not satisfy "latest backup on each hour for 3 hours" -- they are within 3 hours. I would have expected backups at 20:38 and 16:32 instead. Perhaps, I am confused by the meaning "for 7 days / 3 hours"?

To me, it looks like the retention policy for --keep-hourly is:

1. Sort backups by timestamps
2. Group by year,month,day, then hour
3. Mark most recent backup in each hour as a keep candidate
4. Keep only most recent N such candidates, prune the rest

Is that correct?

**Comment by ThomasWaldmann** [comment]
Yes, you've described the algorithm correctly I think.

The "for N timeunits" might trigger wrong expectations, while it is not incorrect considering the implementation.

10:59 falls into the 10h group, 11:42 falls into the 11h group, 12:00 falls into the 12h group.
So while you have covered 3 groups (3 hours), the time difference between the first and the last is only 1 hour and one minute.

The good news is that covering ~N-2 units is the worst case possible (being very late in the first group and extremely early in the last group).

Also, if you run your backups in a cron job always at the same minute-of-the-hour or hour-of-the-day, this won't happen.

So, I don't think there is anything to fix in the code. Can the docs / help get improved?

**Comment by losvald** [comment]
Long-term docs/help improvement would be adding examples similar as "borg help patterns", but for now I suggest rewording the misleading explanation of "-d 7" to something like this:

> "-d 7" keeps the latest backup on each day, up to 7 most recent days with backups".

**Comment by losvald** [comment]
Long-term docs/help improvement would be adding examples similar as "borg help patterns", but for now I suggest rewording the misleading explanation of "-d 7" to something like this:

> "-d 7" keeps the latest backup on each day, up to 7 most recent days with backups".

**Comment by jdchristensen** [comment]
@losvald Have you looked at the output of the "help prune" command?  I wrote that help text precisely because I had trouble understanding how pruning works, so the intention is that it should be very clear.  Feel free to suggest improvements.

$ borg help prune
usage: borg prune [-h] [-v] [--debug] [--lock-wait N] [--show-rc]
                          [--no-files-cache] [--umask M] [--remote-path PATH]
                          [-n] [-s] [--keep-within WITHIN] [-H HOURLY]
                          [-d DAILY] [-w WEEKLY] [-m MONTHLY] [-y YEARLY]
                          [-p PREFIX] [--save-space]
                          [REPOSITORY]

Prune repository archives according to specified rules

positional arguments:
  REPOSITORY            repository to prune

optional arguments:
  -h, --help            show this help message and exit
  -v, --verbose, --info
                        enable informative (verbose) output, work on log level
                        INFO
  --debug               enable debug output, work on log level DEBUG
  --lock-wait N         wait for the lock, but max. N seconds (default: 1).
  --show-rc             show/log the return code (rc)
  --no-files-cache      do not load/update the file metadata cache used to
                        detect unchanged files
  --umask M             set umask to M (local and remote, default: 0077)
  --remote-path PATH    set remote path to executable (default: "borg")
  -n, --dry-run         do not change repository
  -s, --stats           print statistics for the deleted archive
  --keep-within WITHIN  keep all archives within this time interval
  -H HOURLY, --keep-hourly HOURLY
                        number of hourly archives to keep
  -d DAILY, --keep-daily DAILY
                        number of daily archives to keep
  -w WEEKLY, --keep-weekly WEEKLY
                        number of weekly archives to keep
  -m MONTHLY, --keep-monthly MONTHLY
                        number of monthly archives to keep
  -y YEARLY, --keep-yearly YEARLY
                        number of yearly archives to keep
  -p PREFIX, --prefix PREFIX
                        only consider archive names starting with this prefix
  --save-space          work slower, but using less space

The prune command prunes a repository by deleting archives not matching
any of the specified retention options. This command is normally used by
automated backup scripts wanting to keep a certain number of historic backups.

As an example, "-d 7" means to keep the latest backup on each day for 7 days.
Days without backups do not count towards the total.
The rules are applied from hourly to yearly, and backups selected by previous
rules do not count towards those of later rules. The time that each backup
completes is used for pruning purposes. Dates and times are interpreted in
the local timezone, and weeks go from Monday to Sunday. Specifying a
negative number of archives to keep means that there is no limit.

The "--keep-within" option takes an argument of the form "<int><char>",
where char is "H", "d", "w", "m", "y". For example, "--keep-within 2d" means
to keep all archives that were created within the past 48 hours.
"1m" is taken to mean "31d". The archives kept with this option do not
count towards the totals specified by any other options.

If a prefix is set with -p, then only archives that start with the prefix are
considered for deletion and only those archives count towards the totals
specified by the rules.
Otherwise, *all* archives in the repository are candidates for deletion!

**Comment by ThomasWaldmann** [comment]
Fixed by https://github.com/borgbackup/borg/commit/695dc684794daa47152a8182b3855c0f08992d60

**Comment by jdchristensen** [comment]
The revised version has a run-on sentence.  How about:

As an example, "-d 7" means to keep the latest backup on each day, for the 7
most recent days with backups (days without backups do not count).
